### PR TITLE
feat: introduce evm config trait

### LINKED ISSUE #6180: Add Stateful EVM component trait
### Describe the feature

#6051 adds a stateless config trait, though in order to support things like custom precompiles we need to capture state, e.g.

```rust
trait EvmConfig: EvmEnvConfig {
  fn precompiles(&self) -> Precompiles;

  // this is also where all the handlers will be added with EvmBuilder
}
```

this should also abstract away the `EvmProcessorFactory`

https://github.com/paradigmxyz/reth/blob/9a5120a8834a9355d1ec9ad979286d836d747a85/crates/revm/src/factory.rs#L12-L12

perhaps we try:

```rust
  type Executor : PrunableBlockExecutor;
  fn executor(&self) -> Self::Executor;
```

basically this function:

https://github.com/paradigmxyz/reth/blob/9a5120a8834a9355d1ec9ad979286d836d747a85/crates/storage/provider/src/traits/executor.rs#L14-L14

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
closes #6180

### CODE REVIEW

File: crates/node-api/src/evm/traits.rs
Comment (onbjerg) [inline crates/node-api/src/evm/traits.rs:58]: do we dedupe with https://github.com/paradigmxyz/reth/blob/2a5efb2a99ca50a45e41c50b5ef0e6d838d46b28/crates/storage/provider/src/traits/executor.rs#L21 eventually?

Comment (Rjected) [reply crates/node-api/src/evm/traits.rs:58]: yep, actually I think I'll move this to a new crate that `reth-revm` and `node-api` will both depend on, instead of including this (and relevant types) in `node-api`. Mainly to avoid circular dependencies, and avoid making `node-api` gigantic

Comment (Rjected) [reply crates/node-api/src/evm/traits.rs:58]: actually, hmm, need to think more about this, this might not work

**Comment by Rjected** [comment]
final challenge w.r.t. the BlockExecutor trait: `BundleStateWithReceipts`, part of the `BlockExecutor` trait, is currently in `reth-provider` and has a method `write_to_db`. This pulls in methods from `reth-db` (for writing) and pulls in the `reth-provider` types `StateChanges` and `StateReverts`. These types wrap corresponding revm types and implement a `write_to_db` method.

Two potential solutions:
* We could move the non-db-related `BundleStateWithReceipts` methods to `node-api`, and wrap it, similar to `StateChanges` and `StateReverts`.
* Create provider traits responsible for reading and writing these data structures. For example, `BlockWriter` was created to replace the same methods in `Block`: https://github.com/paradigmxyz/reth/pull/3384

Wrapping is easy but the trait seems more idiomatic, so I'm proposing we create a trait that can write `BundleStateWithReceipts`, `StateChanges`, and `StateReverts`

**Comment by Rjected** [comment]
Defining `BlockExecutor` in `node-api` for this purpose:
```rust
trait EvmConfig {
    /// the executor type
    type Executor: BlockExecutor;
    
    /// returns the executor with the installed db
    fn evm(&self, db: impl Database) -> Self::Executor;
}
```
has been scoped out for a few reasons:
1. EvmProcessor has a lifetime, which means using `BlockExecutor` as a bound on an associated type would require that associated type to be a GAT.
2. The alternative, returning `EvmProcessor<'static>` would require the DB to be `'static`, which is basically never true.

The idea to have `take_output_state` consume self was proposed in #6568, but is basically impossible, unless we want to run into the above problems with GATs and object safety because `ExecutorFactory` returns a `Box<dyn PrunableBlockExecutor>`, which we can't turn into an owned value

### Where are we now?

Now we're focusing on a trait like this:
```rust
/// Trait for configuring the EVM for executing full blocks.
pub trait EvmConfig {
    /// Returns new EVM builder
    fn evm(&self) -> EvmBuilder<'static, SetGenericStage, (), EmptyDB> {
        Default::default()
    }
}
```

File: crates/node-optimism/src/evm.rs
Comment (mattsse) [inline crates/node-optimism/src/evm.rs:51]: this needs &self

**Review by mattsse — COMMENTED** [review]
we need them to be stateful, so we can capture additional state for precompiles etc.

File: crates/node-optimism/src/evm.rs
Comment (Rjected) [reply crates/node-optimism/src/evm.rs]: just added

File: crates/node-api/src/evm/traits.rs
Comment (onbjerg) [inline crates/node-api/src/evm/traits.rs:21]: nit: add docs on what transaction metadata is in this context (e.g. give an example)

File: crates/node-api/src/evm/traits.rs
Comment (onbjerg) [inline crates/node-api/src/evm/traits.rs:18]: nit: make it a bit more distinct from `ConfigureEvm` docs

```suggestion
/// This represents the set of methods used to configure the EVM's environment before block execution.
```

File: crates/node-optimism/src/evm.rs
Comment (onbjerg) [inline crates/node-optimism/src/evm.rs:50]: nit: technically don't need to dupe the docs here (and below)

**Review by onbjerg — CHANGES_REQUESTED** [review]
small nits

**Review by mattsse — COMMENTED** [review]
This should unblock us.

then we can figure out how to make the blockexecutor work.

feel free to close this PR and submit it as your own, can't approve myself.

**Review by onbjerg — APPROVED** [review]

**Comment by Rjected** [comment]
@mattsse handler method is blocked on https://github.com/bluealloy/revm/pull/1124 - going to merge this, so we can work on the handler method and evm processor refactor separately

File: crates/node-api/src/evm/traits.rs
Comment (Rjected) [reply crates/node-api/src/evm/traits.rs:12]: ah forgot to push these changes before merging, will add in another PR

**Pull request merged**
